Code/GraphMol/Sgroup.cpp

```diff
@@ -0,0 +1,322 @@
+//
+//
+//  Copyright (C) 2002-2018 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include "Sgroup.h"
+#include "ROMol.h"
+
+namespace RDKit {
+
+static const std::vector<std::string> sGroupTypes = {
+    // polymer sgroups:
+    "SRU", "MON", "COP", "CRO", "GRA", "MOD", "MER", "ANY",
+    // formulations/mixtures:
+    "COM", "MIX", "FOR",
+    // other
+    "SUP", "MUL", "DAT", "GEN"};
+
+static const std::vector<std::string> sGroupSubtypes = {"ALT", "RAN", "BLO"};
+static const std::vector<std::string> sGroupConnectTypes = {"HH", "HT", "EU"};
+
+SGroup::SGroup(ROMol *owning_mol, const std::string &_type)
+    : d_id(0), d_type(_type), dp_mol(owning_mol) {
+  PRECONDITION(owning_mol, "owning molecule is nullptr");
```

Comment:
A friendlier comment is SGroup has no owning molecule.

Reply:
Hm. But that is not strictly true: the SGroup doesn't exist yet (we are still constructing it), and we are checking that the molecule that should own it does really exist.

Maybe "owning molecule does not exist"?

Reply:
My 2cents, you are making an interface that takes a pointer but doesn't except a null ptr.  One way to remove this issue is by only taking a reference.

```
SGroup::SGroup(ROMol &owning_mol, ...):
  dp_mol(&owning_mol), 
```

This removes the issue (and adds others admittedly)

As for the message:

Perhaps, "supplied owning molecule is null" or "supplied owning molecule is bad" to use the "bad molecule" term from the rest of rdkit :)  One thing I tend to have in the back of my head is that these messages will be propagated to the wrapped languages, so I try to make them a bit more general, but it isn't crucial by any means.